openmdao/core/component.py

```diff
@@ -829,3 +829,39 @@ def complex_step_jacobian(self, params, unknowns, resids, total_derivs=False,
             stepvec.set_complex_var(None)
 
         return jac
+
+    def alloc_jacobian(self):
+        """
+        Creates a jacobian dictionary with the keys pre-populated and correct
+        array sizes allocated. caches the result in the component, and
+        returns that cache if it finds it.
+
+        Returns
+        -----------
+        dict
+            pre-allocated jacobian dictionary
+        """
+
+        if self._jacobian_cache is not None and len(self._jacobian_cache) > 0:
+            return self._jacobian_cache
+
+        self._jacobian_cache = jac = {}
+
+        u_vec = self.unknowns
+        p_vec = self.params
+        states = self.states
+
+        for u_var in iterkeys(u_vec):
+            meta = u_vec.metadata(u_var)
+            u_size = meta['size']
+
+            for p_var in iterkeys(p_vec):
+                p_size = p_vec.metadata(p_var)['size']
```

Comment:
I'm not sure how much difference it would make since this is a cached result and is only done once, but it could help setup time for some of these models with tons of variables if you move the calculation of p_size and s_size out of this loop, since you only need to do those lookups p times and s times instead of p x u times and s x u times.